examples/pytorch_ignite/gin.py

```diff
@@ -0,0 +1,214 @@
+        pin_memory=torch.cuda.is_available(),
+    )
+    loader_val = pyg.data.DataLoader(
+        dataset=val_dataset,
+        batch_size=batch_size,
+        shuffle=False,
+        num_workers=1,
+        pin_memory=torch.cuda.is_available(),
+    )
+    loader_test = pyg.data.DataLoader(
+        dataset=test_dataset,
+        batch_size=batch_size,
+        shuffle=False,
+        num_workers=1,
+        pin_memory=torch.cuda.is_available(),
+    )
+
+    model = Model(dataset.num_node_features, dataset.num_classes).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+    loss = torch.nn.functional.cross_entropy
+    metrics = {'acc': ignite.metrics.Accuracy()}
+
+    def prepare_batch_fn(batch, device, non_blocking):
+        # return a tuple of (x, y) that can be directly runned as
+        # `loss_fn(model(x), y)`
+        # dato che ignite non dà argomenti multipli a model, dovremo
+        # avere modelli che estraggono dentro forward() i dati di input
+        # da `data`
+        return (batch.to(device, non_blocking=non_blocking),
+                batch.y.to(device, non_blocking=non_blocking))
+
+    trainer = ignite.engine.create_supervised_trainer(
+        model=model,
+        optimizer=optimizer,
+        loss_fn=loss,
+        device=device,
+        prepare_batch=prepare_batch_fn,
+        output_transform=lambda x, y, y_pred, loss: loss.item(),
+        amp_mode=amp_mode,
+    )
+    # progress bar for each epoch
+    pbar = ignite.contrib.handlers.tqdm_logger.ProgressBar()
+    pbar.attach(trainer, output_transform=lambda x: {"loss": x})
+
+    trn_evaluator = ignite.engine.create_supervised_evaluator(
+        model=model, metrics=metrics, device=device,
+        prepare_batch=prepare_batch_fn, output_transform=lambda x, y, y_pred:
+        (y_pred, y), amp_mode=amp_mode)
+
+    @trainer.on(ignite.engine.Events.EPOCH_COMPLETED(every=1))
+    def log_training_metrics(trainer):
+        trn_evaluator.run(loader_trn)
+        metrics = trn_evaluator.state.metrics
+        print(f"{'Training Results - Epoch: ':30}{trainer.state.epoch} ",
+              end='')
+        for key, value in metrics.items():
+            print(f"{key}: {value:.2f} ", end='')
+        print()
+
+    val_evaluator = ignite.engine.create_supervised_evaluator(
+        model=model,
+        metrics=metrics,
+        device=device,
+        prepare_batch=prepare_batch_fn,
+        output_transform=lambda x, y, y_pred: (y_pred, y),
+        amp_mode=amp_mode,
+    )
+
+    @trainer.on(ignite.engine.Events.EPOCH_COMPLETED(every=1))
+    def log_validation_metrics(trainer):
+        val_evaluator.run(loader_val)
+        metrics = val_evaluator.state.metrics
+        print(f"{'Validation Results - Epoch: ':30}{trainer.state.epoch} ",
+              end='')
+        for key, value in metrics.items():
+            print(f"{key}: {value:.2f} ", end='')
+        print()
+
+    test_evaluator = ignite.engine.create_supervised_evaluator(
```

Comment:
I have to create three different evaluators (even if they are equal) because, after creation, I attach different functions for personalized logging to each one. Don't know if there's a better solution to this.

Reply:
Can't we generate a single function `log_metrics` and call it within `log_training_metrics`, etc?